### PR TITLE
PLG-108: Updated product view params according to new api spec

### PR DESCRIPTION
**Pull request by nedick007**
Refactoring product_view event to be eligible with new api specs

### CODE REVIEW

**Comment by nedick007** [comment]
@adamof view_product event is ready for review

**Comment by nedick007** [comment]
@adamof deprecated registry (in 2.3) i mentioned yesterday can be a subject of discussion. This article and the comment section can clarify the case https://www.atwix.com/magento-2/alternatives-for-deprecated-registry-class-magento-2-3/

File: view/frontend/templates/page/analytics.phtml [hunk 1]
```diff
@@ -7,25 +7,16 @@
...
-        metrilo.identify("<?php echo $event['data']['id']; ?>", <?php echo json_encode($event['data']['params']); ?>);
-        <?php elseif($event['method'] == 'track') : ?>
-            <?php if(is_array($event['data'])): ?>
-                metrilo.event("<?php echo $event['type']; ?>", <?php echo json_encode($event['data']); ?>);
-            <?php else : ?>
-                <?php if(array_key_exists('metaData', $event)) : ?>
-                    metrilo.event("<?php echo $event['type']; ?>", "<?php echo $event['data']; ?>", <?php echo json_encode($event['metaData']); ?>);
-                <?php else : ?>
-                    metrilo.event("<?php echo $event['type']; ?>", "<?php echo $event['data']; ?>");
+    require(['<?php echo $block->getApiEndpoint(); ?>' + '/tracking.js?token=' + '<?php echo $block->getApiToken(); ?>'], function(){
```
Comment (adamof) [inline view/frontend/templates/page/analytics.phtml]: Please build this URL in the PHP block on top to make it more readable

File: view/frontend/templates/page/analytics.phtml [hunk 2]
```diff
@@ -7,25 +7,16 @@
...
+        <?php foreach ($events as $event) : ?>
+            console.log('<?php echo 'Event Type: ' . $event['type']; ?>');
+            <?php if($event['method'] == 'identify') : ?>
```
Comment (adamof) [inline view/frontend/templates/page/analytics.phtml]: remove

File: view/frontend/templates/page/analytics.phtml (hunk 1 above)
Comment (nedick007) [reply view/frontend/templates/page/analytics.phtml]: Refactored.

File: view/frontend/templates/page/analytics.phtml (hunk 2 above)
Comment (nedick007) [reply view/frontend/templates/page/analytics.phtml]: Removed.

**Comment by nedick007** [comment]
@adamof old events are removed.

File: view/frontend/templates/page/analytics.phtml [hunk 3]
```diff
@@ -5,27 +5,17 @@
...
-                metrilo.event("<?php echo $event['type']; ?>", <?php echo json_encode($event['data']); ?>);
-            <?php else : ?>
-                <?php if(array_key_exists('metaData', $event)) : ?>
-                    metrilo.event("<?php echo $event['type']; ?>", "<?php echo $event['data']; ?>", <?php echo json_encode($event['metaData']); ?>);
-                <?php else : ?>
-                    metrilo.event("<?php echo $event['type']; ?>", "<?php echo $event['data']; ?>");
+    require(['<?php echo $url; ?>'], function(){
+        <?php foreach ($events as $event) : ?>
+            console.log('<?php echo 'Event Type: ' . $event['type']; ?>');
+            <?php if($event['method'] == 'track') : ?>
```
Comment (adamof) [inline view/frontend/templates/page/analytics.phtml]: Is there any other type of events?

File: Model/Analytics.php [hunk 4]
```diff
@@ -55,84 +49,21 @@ public function addPageEvents()
...
-                'id'    => $product->getId(),
-                'sku'   => $product->getSku(),
-                'name'  => $product->getName(),
-                'price' => $product->getPrice(),
-                'url'   => $product->getProductUrl()
+        
+        switch($this->fullActionName) {
+            // product view pages
+            case 'catalog_product_view':
+                /** @var \Magento\Catalog\Model\Product $product */
```
Comment (adamof) [inline Model/Analytics.php]: Here:
```php
event = ProductViewEvent.new()
$this->addEvent('track', 'view_product', $data);
```
ProductViewEvent.php
```php
public function __construct(
    \Magento\Framework\Registry $registry
) {
    $this->_coreRegistry = $registry;
}
public function callJS() {
    $product = $this->_coreRegistry->registry('current_product');
    return "window.metrilo.viewProduct(product.id);";
}
```

File: view/frontend/templates/page/analytics.phtml [hunk 5]
```diff
@@ -5,27 +5,17 @@
...
-        <?php elseif($event['method'] == 'track') : ?>
-            <?php if(is_array($event['data'])): ?>
-                metrilo.event("<?php echo $event['type']; ?>", <?php echo json_encode($event['data']); ?>);
-            <?php else : ?>
-                <?php if(array_key_exists('metaData', $event)) : ?>
-                    metrilo.event("<?php echo $event['type']; ?>", "<?php echo $event['data']; ?>", <?php echo json_encode($event['metaData']); ?>);
-                <?php else : ?>
-                    metrilo.event("<?php echo $event['type']; ?>", "<?php echo $event['data']; ?>");
+    require(['<?php echo $url; ?>'], function(){
+        <?php foreach ($events as $event) : ?>
```
Comment (adamof) [inline view/frontend/templates/page/analytics.phtml]:
```
<?php echo $event.callJS; ?>
```

File: Model/Analytics.php (hunk 4 above)
Comment (nedick007) [reply Model/Analytics.php]: Removed Model/Analytics and moved event logic to the Block/Analytics class.

File: view/frontend/templates/page/analytics.phtml (hunk 5 above)
Comment (nedick007) [reply view/frontend/templates/page/analytics.phtml]: Refactored.

**Comment by nedick007** [comment]
@adamof bump

File: Helper/Events/ProductViewEvent.php [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+class ProductViewEvent extends \Magento\Framework\App\Helper\AbstractHelper
+{
+    public function __construct(
+        \Magento\Framework\Registry $registry
+    ) {
+        $this->_coreRegistry = $registry;
+    }
+    public function callJS() {
+        $product = $this->_coreRegistry->registry('current_product');
```
Comment (adamof) [inline Helper/Events/ProductViewEvent.php]: Can we optimise the query here to return only the product ID?

File: Block/Analytics.php [hunk 7]
```diff
@@ -19,22 +19,24 @@ class Analytics extends Template
...
+     * @param array                                             $data
      */
     public function __construct(
         Context $context,
-        \Metrilo\Analytics\Helper\Data $helper,
-        \Metrilo\Analytics\Model\Analytics $dataModel,
-        \Magento\Customer\Model\Session $session,
+        \Magento\Framework\App\Action\Context             $actionContext,
+        \Metrilo\Analytics\Helper\Data                    $helper,
+        \Metrilo\Analytics\Helper\Events\ProductViewEvent $productViewEvent,
```
Comment (adamof) [inline Block/Analytics.php]: Do we really want this to be a helper? We're going to have to add ALL types of events here then. Maybe a model is better? What do you think?

File: Block/Analytics.php [hunk 8]
```diff
@@ -74,4 +86,38 @@ protected function _toHtml()
...
+
+    /**
+     * Events that we don't want to track
+     *
+     * @param string full action name
+     */
+    protected function isRejected($action)
+    {
+        $rejected = [
+            'catalogsearch_advanced_index',
```
Comment (adamof) [inline Block/Analytics.php]: Add a story to think about these. Why are they being ignored?

File: view/frontend/templates/page/analytics.phtml [hunk 9]
```diff
@@ -2,30 +2,11 @@
 /**
  * @see Metrilo\Analytics\Block\Analytics
  */
-
-/** @var Metrilo\Analytics\Block\Analytics $events */
-$events = $block->getEvents();
+$event = $block->getEvent();
+$url   = $block->getApiEndpoint() . '/tracking.js?token=' . $block->getApiToken();
```
Comment (adamof) [inline view/frontend/templates/page/analytics.phtml]: Looking at this, I think it makes sense that the whole URL is returned by either the block. What do you think?

**Comment by nedick007** [comment]
@adamof ready for review

File: Block/Analytics.php [hunk 10]
```diff
@@ -19,33 +19,28 @@ class Analytics extends Template
...
-     * @param Context                            $context
-     * @param \Metrilo\Analytics\Helper\Data     $helper
-     * @param \Metrilo\Analytics\Model\Analytics $dataModel
-     * @param \Magento\Customer\Model\Session    $session
-     * @param array                              $data
+     * @param Context                                           $context
+     * @param \Magento\Framework\App\Action\Context             $actionContext
+     * @param \Metrilo\Analytics\Helper\Data                    $helper
+     * @param \Metrilo\Analytics\Helper\Events\ProductViewEvent $productViewEvent,
+     * @param array                                             $data
```
Comment (adamof) [inline Block/Analytics.php]: 🔥

File: Block/Analytics.php [hunk 11]
```diff
@@ -19,33 +19,28 @@ class Analytics extends Template
...
      */
     public function __construct(
         Context $context,
-        \Metrilo\Analytics\Helper\Data $helper,
-        \Metrilo\Analytics\Model\Analytics $dataModel,
-        \Magento\Customer\Model\Session $session,
+        \Magento\Framework\App\Action\Context            $actionContext,
+        \Metrilo\Analytics\Helper\Data                   $helper,
+        \Metrilo\Analytics\Model\Events\ProductViewEvent $productViewEvent,
```
Comment (adamof) [inline Block/Analytics.php]: 🔥

File: Block/Analytics.php (hunk 8 above)
Comment (nedick007) [reply Block/Analytics.php]: Story url: https://metrilojira.atlassian.net/browse/PLG-110

**Comment by nedick007** [comment]
@adamof bump

File: Block/Analytics.php [hunk 12]
```diff
@@ -61,17 +40,35 @@ public function getEvents()
...
+    public function getEvent()
+    {
+        if (!$this->fullActionName || $this->isRejected($this->fullActionName)) {
+            return;
+        }
+        
+        switch($this->fullActionName) {
+            // product view pages
+            case 'catalog_product_view':
+                return new \Metrilo\Analytics\Model\Events\ProductViewEvent($this->coreRegistry->registry('current_product')->getId());
```
Comment (adamof) [inline Block/Analytics.php]: You're spreading the knowledge of how to build the event in here. All the knowledge should be in the `ProductViewEvent` class. What do you think about passing the registry and letting the ProductViewEvent know what to take out from it?

Comment (nedick007) [reply Block/Analytics.php]: It sounds more reasonable, will fix it.

**Comment by nedick007** [comment]
@adamof bump review

**Comment by nedick007** [comment]
Tested on QA env.

**Pull request merged**